src/net/ossl.cc

```diff
@@ -590,249 +640,446 @@ class session : public enable_shared_from_this<session>, public session_impl {
+                             msg->append(std::move(buf));
+                         } else if (!BIO_should_retry(_out_bio)) {
+                             _error = std::make_exception_ptr(
+                               ossl_error::make_ossl_error(
+                                 "Failed to read data from _out_bio"));
+                             return make_exception_future(_error);
+                         }
+                         return make_ready_future<>();
+                     })
+              .then([this, msg] {
+                  if (msg->size() > 0) {
+                      _output_pending = _out.put(std::move(*msg).release());
+                  } else {
+                      _output_pending = make_ready_future();
+                  }
+                  return make_ready_future<>();
+              });
         });
     }
 
-    // This method puts unencrypted data is written into the SSL context.
-    // This data is later able to be retrieved in its encrypted form by reading
-    // from the associated _out_bio
+    // This function will check to see if there is any data sitting in the
+    // _out_bio, which is the BIO that the SSL session writes to to send
+    // data.  If there is, it call `perform_push` and wait for the data
+    // to be sent.  If there is no data to be sent, this function returns
+    // immediately.
+    // Returns true if data is sent, false if not
+    future<bool> maybe_perform_push_with_wait() {
+        if (BIO_ctrl_pending(_out_bio) > 0) {
+            return perform_push().then([this] {
+                return wait_for_output();
+            }).then([]() {
+                return make_ready_future<bool>(true);
+            });
+        } else {
+            return make_ready_future<bool>(false);
+        }
+    }
+
+    // Helper function for handling the SSL errors in do_put
+    future<stop_iteration> handle_do_put_ssl_err(const int ssl_err) {
+        switch(ssl_err) {
+        case SSL_ERROR_ZERO_RETURN:
+            // Indicates a hang up somewhere
+            // Mark _eof and stop iteratio
+            _eof = true;
+            return make_ready_future<stop_iteration>(stop_iteration::yes);
+        case SSL_ERROR_NONE:
+            // Should not have been reached in this situation
+            // Continue iteration
+            return make_ready_future<stop_iteration>(stop_iteration::no);
+        case SSL_ERROR_WANT_READ:
+        case SSL_ERROR_WANT_WRITE:
+            // These 'error' codes indicate to the caller that before we can
+            // continue writing data to the SSL session, the SSL session needs
+            // to send or receive data from the peer.  Could indicate a
+            // renegotitation is required.
+            return maybe_perform_push_with_wait().then([this](bool) {
+                // The _out_sem semaphore is currently being held, however if we need
+                // to receive data, we need to obtain a lock on the _in_sem semaphore
```

Comment:
Usually (for example, during shutdown and close), in is taken before out. Can taking them in an alternative order here cause a deadlock?

Reply:
yeah i went ahead and pulled those additional semaphore locks out... I think I need them but where I put them was wrong... need to think on it some more